### PR TITLE
feat: pay out yield of PoX to users eligible for yield in stacking

### PR DESCRIPTION
**Pull request by philipdesmedt**
Changes:

- Fixed a bug in revoking stacking where the stacked tokens count in the STX reserve was updated wrong
- Implemented payout of vault owner, or of auction bidders in case of liquidation

### CODE REVIEW

File: clarity/contracts/stacker.clar [hunk 1]
```diff
@@ -19,9 +19,15 @@
 (define-constant ERR-NOT-AUTHORIZED u19401)
 (define-constant ERR-CANNOT-STACK u191)
 (define-constant ERR-FAILED-STACK-STX u192)
+(define-constant ERR-BURN-HEIGHT-NOT-REACHED u193)
+(define-constant ERR-ALREADY-STACKING u194)
 (define-constant CONTRACT-OWNER tx-sender)
```
Comment (nieldeckx) [inline clarity/contracts/stacker.clar:24]: NTH: Is the deployer always the one calling "set-stacking-stx-received"? If a community member updates this contract through governance, he will become the deployer. But it's up to the community member to update this code properly I guess.

File: clarity/contracts/stacker.clar [hunk 2]
```diff
@@ -77,23 +94,105 @@
...
+(define-private (payout-vault (vault-id uint))
+  (let (
+    (vault (contract-call? .vault-data get-vault-by-id vault-id))
+    (earned-amount (calculate-vault-reward vault-id))
+    (new-collateral-amount (+ earned-amount (get collateral vault)))
+  )
+    (asserts! (is-eq (get is-liquidated vault) false) (err ERR-NOT-AUTHORIZED))
+    (asserts! (> (get stacked-tokens vault) u0) (err ERR-NOT-AUTHORIZED))
+
+    (try! (contract-call? .vault-data update-vault vault-id (merge vault { collateral: new-collateral-amount })))
```
Comment (nieldeckx) [inline clarity/contracts/stacker.clar:182]: Q: We collect BTC and swap it to STX. Where do we send the STX to? A payout increases the collateral, so we need to make sure the STX end up in the right contract?

File: clarity/contracts/stacker.clar (hunk 1 above)
Comment (philipdesmedt) [reply clarity/contracts/stacker.clar:24]: Yeah - good remark. My idea is to change CONTRACT-OWNER for something that the DAO can control maybe.. I have a card for this on Trello: https://trello.com/c/wTzlV9pk/74-have-a-guardian-extra-private-key-to-derisk-some-variables-eg-contract-owner-checks - we should discuss an approach to this and can then refactor all these occurrences in a separate PR

File: clarity/contracts/stacker.clar (hunk 2 above)
Comment (philipdesmedt) [reply clarity/contracts/stacker.clar:182]: The STX needs to be sent to the principal of the current implementation of stacker.clar that we keep in the DAO, e.g. 'ST<ADDRESS>.stacker.

File: clarity/contracts/stacker.clar [hunk 3]
```diff
@@ -77,23 +94,104 @@
...
+    (if (get revoked-stacking vault)
+      (begin
+        (try! (contract-call? .vault-data update-vault vault-id (merge vault { updated-at-block-height: block-height, stacked-tokens: u0 })))
+        (try! (request-stx-for-withdrawal new-collateral-amount))
```
Comment (philipdesmedt) [inline clarity/contracts/stacker.clar:185]: @nieldeckx here is the relevant line where the transfer happens back into the `stx-reserve`. The flow is as follows:

1. User creates vault with STX collateral. Funds are kept in the `stx-reserve` smart contract and a counter of who wants to stack is kept.
2. At the start of a stacking cycle, the `stacker` contract requests the STX tokens it needs to stack from the `stx-reserve`
3. If a vault owner changes their mind and does not want to stack anymore, they revoke stacking on their vault
4. At the end of the cycle, this payout method looks at a vault and checks if they revoked stacking (see `if` statement on line 182) and returns their STX tokens to the vault (`stx-reserve` contract) if so

**Pull request merged**